clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
             int totalBytes = 0;
             int totalCount = 0;
             FetchResponse response = new FetchResponse(resp.responseBody());
+
+            // Look for OffsetOutOfRange error in fetchResponse
+            Map<TopicPartition, Long> offsetOutOfRangePartitions = new HashMap();
+            for (Map.Entry<TopicPartition, FetchResponse.PartitionData> entry : response.responseData().entrySet()) {
```

Comment:
I think I am preferring the existing solution more -- it will be easier to say that poll() will always request data for all partitions starting from the offset before poll() is called, and it either throw exception, or return data for all partitions that have new data after the current offset.

The proposal you suggested makes the API harder to explain -- it requires user to know whether there is an OffsetOutOfRange exception in the previous poll() call.